# Atividade Prática 2: Avaliação de classificadores baseados em features - Grupo 6

Participantes: Antonio Henrique, Pedro Lucca, Samuel Santos e Naum Celestino

### Introdução

Este trabalho tem como objetivo a análise textual das contratações emergenciais dos munícipios e Estado do Piauí, realizadas no período de março a setembro de 2020, referentes ao enfrentamento da pandemia da COVID-19. Os dados foram extraídos dos respectivos diários oficiais por uma comissão do Tribunal de Contas do Estado do Piauí, criada especificamente para o acompanhamento dessas contratações, e inseridos em uma planilha Excel. 

Desta forma, o corpus do trabalho foi construído baseado a partir da coluna “Objeto” da planilha em questão, e assim, será analisado o uso das palavras no corpus, extraindo as principais features que nos ajudam a entender o contexto e avaliar o desempenho de classificadores e implementar um novo classificador baseado em features, considerando os dados do corpus utilizado pelo grupo na atividade prática 1.




#### Questão 1

Considerando os classificadores de gêneros de nomes discutidos nas aulas: classif1 e classif2, responder as seguintes perguntas: 

a. Qual o desempenho (acurácia) dos dois modelos, ao utilizar os mesmos
conjuntos de treinamento e testes para o processo de treinamento e avaliação?


In [ ]:
def gender_features(word):
    return {'last_letter': word[-1]}
gender_features('Shrek')
#{‘last_letter’: ‘k’} ; dictionary

{'last_letter': 'k'}

In [ ]:
def gender_features2(name):
  feat = {}
  feat['first_letter'] = name[0].lower()
  feat['last_letter'] = name[-1].lower()
  for let in 'abcdefghijklmnopqrstuvwxyz':
    feat['count({})'.format(let)] = name.lower().count(let)
    feat['has({})'.format(let)] = (let in name.lower())
  return feat

gender_features2('John')


{'count(a)': 0,
 'count(b)': 0,
 'count(c)': 0,
 'count(d)': 0,
 'count(e)': 0,
 'count(f)': 0,
 'count(g)': 0,
 'count(h)': 1,
 'count(i)': 0,
 'count(j)': 1,
 'count(k)': 0,
 'count(l)': 0,
 'count(m)': 0,
 'count(n)': 1,
 'count(o)': 1,
 'count(p)': 0,
 'count(q)': 0,
 'count(r)': 0,
 'count(s)': 0,
 'count(t)': 0,
 'count(u)': 0,
 'count(v)': 0,
 'count(w)': 0,
 'count(x)': 0,
 'count(y)': 0,
 'count(z)': 0,
 'first_letter': 'j',
 'has(a)': False,
 'has(b)': False,
 'has(c)': False,
 'has(d)': False,
 'has(e)': False,
 'has(f)': False,
 'has(g)': False,
 'has(h)': True,
 'has(i)': False,
 'has(j)': True,
 'has(k)': False,
 'has(l)': False,
 'has(m)': False,
 'has(n)': True,
 'has(o)': True,
 'has(p)': False,
 'has(q)': False,
 'has(r)': False,
 'has(s)': False,
 'has(t)': False,
 'has(u)': False,
 'has(v)': False,
 'has(w)': False,
 'has(x)': False,
 'has(y)': False,
 'has(z)': False,
 'last_letter': 'n'}

In [ ]:
import nltk
nltk.download('names')
from nltk.corpus import names
nomesRot = ([(name,'masc') for name in names.words('male.txt')] + [(name,'fem') for name in names.words('female.txt')])
import random
print(nomesRot)
print(len(nomesRot))


[nltk_data] Downloading package names to /root/nltk_data...
[nltk_data]   Package names is already up-to-date!
[('Aamir', 'masc'), ('Aaron', 'masc'), ('Abbey', 'masc'), ('Abbie', 'masc'), ('Abbot', 'masc'), ('Abbott', 'masc'), ('Abby', 'masc'), ('Abdel', 'masc'), ('Abdul', 'masc'), ('Abdulkarim', 'masc'), ('Abdullah', 'masc'), ('Abe', 'masc'), ('Abel', 'masc'), ('Abelard', 'masc'), ('Abner', 'masc'), ('Abraham', 'masc'), ('Abram', 'masc'), ('Ace', 'masc'), ('Adair', 'masc'), ('Adam', 'masc'), ('Adams', 'masc'), ('Addie', 'masc'), ('Adger', 'masc'), ('Aditya', 'masc'), ('Adlai', 'masc'), ('Adnan', 'masc'), ('Adolf', 'masc'), ('Adolfo', 'masc'), ('Adolph', 'masc'), ('Adolphe', 'masc'), ('Adolpho', 'masc'), ('Adolphus', 'masc'), ('Adrian', 'masc'), ('Adrick', 'masc'), ('Adrien', 'masc'), ('Agamemnon', 'masc'), ('Aguinaldo', 'masc'), ('Aguste', 'masc'), ('Agustin', 'masc'), ('Aharon', 'masc'), ('Ahmad', 'masc'), ('Ahmed', 'masc'), ('Ahmet', 'masc'), ('Ajai', 'masc'), ('Ajay', 'masc'), ('Al

In [ ]:
random.shuffle(nomesRot)

#fsets = [(gender_features(n), gen) for (n, gen) in nomesRot]

fsets = []
fsets2 = []

for (n, gen) in nomesRot:
  fsets.append([gender_features(n), gen])
  fsets2.append([gender_features2(n), gen])

size = int(0.1 * len(fsets)) #10% para teste

train, test = fsets[size:], fsets[:size]
train2, test2 = fsets2[size:], fsets2[:size]
classif1 = nltk.NaiveBayesClassifier.train(train)
classif2 = nltk.NaiveBayesClassifier.train(train2)

#print(nltk.classify.accuracy(classif1, test))
print("Taxa de acurácia da classificação 1: ",nltk.classify.accuracy(classif1, test))
print("Taxa de acurácia da classificação 2: ",nltk.classify.accuracy(classif2, test2))

Taxa de acurácia da classificação 1:  0.7695214105793451
Taxa de acurácia da classificação 2:  0.7670025188916877


In [ ]:
#classif1.classify(gender_features('Neo'))
print("Classificação de Neo (classif1): ",classif1.classify(gender_features('Neo')))
print("Classificação de Neo (classif2): ",classif2.classify(gender_features('Neo')))
#'masc'

Classificação de Neo (classif1):  masc
Classificação de Neo (classif2):  masc


In [ ]:
#classif1.classify(gender_features('Trinity'))
print("Classificação de Neo (classif1): ",classif1.classify(gender_features('Trinity')))
print("Classificação de Neo (classif2): ",classif2.classify(gender_features('Trinity')))
#'fem'

Classificação de Neo (classif1):  fem
Classificação de Neo (classif2):  fem


In [ ]:
classif1.show_most_informative_features(10)
classif2.show_most_informative_features(10)

Most Informative Features
             last_letter = 'a'               fem : masc   =     34.9 : 1.0
             last_letter = 'k'              masc : fem    =     30.3 : 1.0
             last_letter = 'f'              masc : fem    =     15.0 : 1.0
             last_letter = 'p'              masc : fem    =     11.0 : 1.0
             last_letter = 'v'              masc : fem    =      9.7 : 1.0
             last_letter = 'm'              masc : fem    =      9.7 : 1.0
             last_letter = 'd'              masc : fem    =      9.3 : 1.0
             last_letter = 'o'              masc : fem    =      8.8 : 1.0
             last_letter = 'r'              masc : fem    =      6.7 : 1.0
             last_letter = 'w'              masc : fem    =      6.1 : 1.0
Most Informative Features
             last_letter = 'a'               fem : masc   =     34.9 : 1.0
             last_letter = 'k'              masc : fem    =     30.3 : 1.0
             last_letter = 'f'              masc

b. Qual o desempenho (acurácia) dos dois modelos, ao utilizar a técnica 10-fold cross validation e os mesmos conjuntos de treinamento e testes para o processo de treinamento e avaliação?

Desempenho (acurácia) do modelo ao utilizar a técnica 10-fold
cross validation 
- Dividir o	corpus em N	subsets,	chamados folds

- Para	cada fold,	treinamos um	modelo usando
todos os dados,	exceto os dados	desse fold e	
então testamos o	modelo com	ele.

In [ ]:
import numpy as np
sum_classif1 = 0
sum_classif2 = 0
round_classify = 1

random.shuffle(nomesRot)

#fsets = [(gender_features(n), gen) for (n, gen) in nomesRot]

fsets = []
fsets2 = []

for (n, gen) in nomesRot:
  fsets.append([gender_features(n), gen])
  fsets2.append([gender_features2(n), gen])

size = int(0.1 * len(fsets)) #10% para teste

for j in range(10):
  i = j+1
  if i == 1:
    train, test = fsets[size:], fsets[:size]
    train2, test2 = fsets2[size:], fsets2[:size]
  elif i == 10:
    train, test = fsets[:size], fsets[size:] 
    train2, test2 = fsets2[:size], fsets2[size:]
  else:
    args = (fsets[:size*(i-1)],fsets[size*i:])
    args2 = (fsets2[:size*(i-1)],fsets2[size*i:])
    train, test = np.concatenate(args), fsets[size*(i-1):size*i]
    train2, test2 = np.concatenate(args2), fsets2[size*(i-1):size*i]


  classif1 = nltk.NaiveBayesClassifier.train(train)
  classif2 = nltk.NaiveBayesClassifier.train(train2)

  #print(nltk.classify.accuracy(classif1, test))
  print(f"ROUND: {round_classify} - Taxa de acurácia da classificação 1: ",nltk.classify.accuracy(classif1, test))
  print(f"ROUND: {round_classify} - Taxa de acurácia da classificação 2: ",nltk.classify.accuracy(classif2, test2))

#####
#    train_data = np.array(fsets)[train]
#    test_data = np.array(fsets)[test]
#    classifier = nltk.NaiveBayesClassifier.train(train_data)
#    print(f"ROUND: {round_classify} - Taxa de acurácia da classificação: ",nltk.classify.accuracy(classifier, test_data))
#####

  sum_classif1 += nltk.classify.accuracy(classif1, test)
  sum_classif2 += nltk.classify.accuracy(classif2, test2)
  round_classify += 1

average1 = round(sum_classif1/10,3)
average2 = round(sum_classif2/10,3)

print(f"\nTaxa de acurácia média da classificação 1: {average1}")
print(f"\nTaxa de acurácia média da classificação 2: {average2}")

ROUND: 1 - Taxa de acurácia da classificação 1:  0.7443324937027708
ROUND: 1 - Taxa de acurácia da classificação 2:  0.7493702770780857
ROUND: 2 - Taxa de acurácia da classificação 1:  0.7506297229219143
ROUND: 2 - Taxa de acurácia da classificação 2:  0.7632241813602015
ROUND: 3 - Taxa de acurácia da classificação 1:  0.7380352644836272
ROUND: 3 - Taxa de acurácia da classificação 2:  0.7720403022670025
ROUND: 4 - Taxa de acurácia da classificação 1:  0.7695214105793451
ROUND: 4 - Taxa de acurácia da classificação 2:  0.7632241813602015
ROUND: 5 - Taxa de acurácia da classificação 1:  0.7644836272040302
ROUND: 5 - Taxa de acurácia da classificação 2:  0.77455919395466
ROUND: 6 - Taxa de acurácia da classificação 1:  0.7770780856423174
ROUND: 6 - Taxa de acurácia da classificação 2:  0.7770780856423174
ROUND: 7 - Taxa de acurácia da classificação 1:  0.756926952141058
ROUND: 7 - Taxa de acurácia da classificação 2:  0.783375314861461
ROUND: 8 - Taxa de acurácia da classificação 1:  0.7

#### 2

Modificar a função gender_features() para criar um classificador usando outras
features (ex: tamanho do nome, primeira letra do nome, uso de dígrafos, ...). Avaliar o novo classificador considerando as mesmas condições descritas na questão 1.

In [ ]:
import re
pattern_ditongo = re.compile('[aeiou][aeiou]')
pattern_tritongo = re.compile('[aeiou][aeiou][aeiou]')
pattern_ditongo_end = re.compile('[a-z]*[aeiou][aeiou]')

def gender_features3(name):
  feat = {}
  feat['first_letter'] = name[0].lower()
  feat['last_letter'] = name[-1].lower()
  feat['len_name'] = len(name)
  feat['ditongo'] = False
  if pattern_ditongo.findall(name.lower()):
    feat['ditongo'] = True
  feat['ditongo_end'] = False
  if pattern_ditongo_end.match(name.lower()):
    feat['ditongo_end'] = True
  for let in 'abcdefghijklmnopqrstuvwxyz':
    feat['count({})'.format(let)] = name.lower().count(let)
    feat['has({})'.format(let)] = (let in name.lower())
  return feat

gender_features3('Marcia')


{'count(a)': 2,
 'count(b)': 0,
 'count(c)': 1,
 'count(d)': 0,
 'count(e)': 0,
 'count(f)': 0,
 'count(g)': 0,
 'count(h)': 0,
 'count(i)': 1,
 'count(j)': 0,
 'count(k)': 0,
 'count(l)': 0,
 'count(m)': 1,
 'count(n)': 0,
 'count(o)': 0,
 'count(p)': 0,
 'count(q)': 0,
 'count(r)': 1,
 'count(s)': 0,
 'count(t)': 0,
 'count(u)': 0,
 'count(v)': 0,
 'count(w)': 0,
 'count(x)': 0,
 'count(y)': 0,
 'count(z)': 0,
 'ditongo': True,
 'ditongo_end': True,
 'first_letter': 'm',
 'has(a)': True,
 'has(b)': False,
 'has(c)': True,
 'has(d)': False,
 'has(e)': False,
 'has(f)': False,
 'has(g)': False,
 'has(h)': False,
 'has(i)': True,
 'has(j)': False,
 'has(k)': False,
 'has(l)': False,
 'has(m)': True,
 'has(n)': False,
 'has(o)': False,
 'has(p)': False,
 'has(q)': False,
 'has(r)': True,
 'has(s)': False,
 'has(t)': False,
 'has(u)': False,
 'has(v)': False,
 'has(w)': False,
 'has(x)': False,
 'has(y)': False,
 'has(z)': False,
 'last_letter': 'a',
 'len_name': 6}

In [ ]:
random.shuffle(nomesRot)

fsets = []
fsets2 = []
fsets3 = []

for (n, gen) in nomesRot:
  fsets.append([gender_features(n), gen])
  fsets2.append([gender_features2(n), gen])
  fsets3.append([gender_features3(n), gen])

size = int(0.1 * len(fsets)) #10% para teste

train, test = fsets[size:], fsets[:size]
train2, test2 = fsets2[size:], fsets2[:size]
train3, test3 = fsets3[size:], fsets3[:size]
classif1 = nltk.NaiveBayesClassifier.train(train)
classif2 = nltk.NaiveBayesClassifier.train(train2)
classif3 = nltk.NaiveBayesClassifier.train(train3)

#print(nltk.classify.accuracy(classif1, test))
print("Taxa de acurácia da classificação 1: ",nltk.classify.accuracy(classif1, test))
print("Taxa de acurácia da classificação 2: ",nltk.classify.accuracy(classif2, test2))
print("Taxa de acurácia da classificação 3: ",nltk.classify.accuracy(classif3, test3))

Taxa de acurácia da classificação 1:  0.7695214105793451
Taxa de acurácia da classificação 2:  0.7884130982367759
Taxa de acurácia da classificação 3:  0.7858942065491183


In [ ]:
classif1.show_most_informative_features(10)
classif2.show_most_informative_features(10)
classif3.show_most_informative_features(10)

Most Informative Features
             last_letter = 'a'               fem : masc   =     34.1 : 1.0
             last_letter = 'k'              masc : fem    =     32.0 : 1.0
             last_letter = 'v'              masc : fem    =     18.6 : 1.0
             last_letter = 'p'              masc : fem    =     17.4 : 1.0
             last_letter = 'f'              masc : fem    =     15.9 : 1.0
             last_letter = 'd'              masc : fem    =     10.4 : 1.0
             last_letter = 'm'              masc : fem    =      9.1 : 1.0
             last_letter = 'o'              masc : fem    =      8.4 : 1.0
             last_letter = 'r'              masc : fem    =      6.5 : 1.0
             last_letter = 'g'              masc : fem    =      5.9 : 1.0
Most Informative Features
             last_letter = 'a'               fem : masc   =     34.1 : 1.0
             last_letter = 'k'              masc : fem    =     32.0 : 1.0
             last_letter = 'v'              masc

#### 3. 

Implementar um classificador baseado em features, considerando os dados do corpus utilizado pelo grupo na atividade prática 02


##### Tratamento do Corpus


In [ ]:
#Para a leitura do arquivo da segunda atividade
import io
import pandas as pd
import nltk
from sklearn.feature_extraction.text import CountVectorizer

git_doc = 'https://github.com/PedroLucca/PLN-exercises/blob/master/diarios_oficiais_COVID_classific.xlsx?raw=true'
dfs = pd.read_excel(git_doc, sheet_name=None)
dfs['diarios_oficiais'].head(5)


Esfera  ... Natureza_de_Despesa
0  Estadual  ...       Não informado
1  Estadual  ...       Não informado
2  Estadual  ...       Não informado
3  Estadual  ...       Não informado
4  Estadual  ...       Não informado

[5 rows x 22 columns]

In [ ]:
diarios = dfs['diarios_oficiais']
#corpus = pd.DataFrame(diarios, columns= ['Objeto'])
#obj = corpus.head(5).to_numpy()
corpus = diarios.get('Objeto').to_numpy().tolist()
print(corpus[-5:])
corpus_obj_simp = diarios.get('Objeto_simplificado').to_numpy().tolist()

['Contratação de Empresa para fornecimentos de Teste Rápido para detecção especifica de IGG e IGM do COVID-19, para o Município de Dom Inocêncio - PI.', 'Prestação de contrataçao de profissional para dar suporte técnico ao desenvolvimento do trabalho de: Acompanhamento a Lei Emergenclal Cultural Aldir Biano, sancionada através do Decreto Federal no 14.017/2020; Alimentação dos sistemas e plataformas necessárias; acompanhamento e julgamento dos editais, todo suporte técnico ao Setor de Cultura junto a Secretaria Municipal de Cultura, Esporte e Lazer e a Prefeitura Municipal de São Jose do Piaul - PI.', 'AQUISIÇÃO DE PULVERIZADOR 20 LT FIBRA BICO PLASTICO PARA SECRETARIA MUNICIPAL DE EDUCAÇÃO E ESCOLAS DA REDE ESTADUAL E MUNICIPAL DO MUNICÍPIO DE SÃO JOSÉ DO PIAUÍ· PI.', 'AQUISIÇÃO DE ÁGUA SANITÁRIA I LT, DETERGENTE 500 ML E PAPEL TOALHA PARA SECRETARIA MUNICIPAL DE EDUCAÇÃO E SEUS RESPECTNOS ÓRGÃOS DO MUNICtPIO DE SÃO JOSE DO PIAUÍ - PI.', 'PRESTAÇÃO DE SERVIÇOS DE TROCA DE PISO DA SECR

In [ ]:
from collections import Counter

Counter(diarios.get('Objeto_simplificado'))

Counter({'Aquisição de EPIs e/ou medicamentos': 1431,
         'Aquisição de ambulâncias e/ou veículos': 11,
         'Aquisição de combustíveis e/ou lubrificantes': 22,
         'Aquisição de equipamentos diversos': 200,
         'Aquisição de equipamentos médico-hospitalares': 283,
         'Aquisição de insumos laboratoriais e/ou material médico-hospitalar': 347,
         'Aquisição de materiais de distribuição gratuita': 15,
         'Aquisição de materiais diversos': 223,
         'Aquisição de material de limpeza': 98,
         'Aquisição de testes para diagnóstico de COVID-19': 326,
         'Construção, reforma e/ou ampliação de unidade hospitalar': 76,
         'Contratação de bens e serviços diversos': 259,
         'Contratação de entidade para gestão hospitalar': 8,
         'Contratação de prestador de serviço - pessoa física': 2283,
         'Contratação de prestador de serviço - pessoa jurÍdica': 5,
         'Contratação de prestador de serviço - pessoa jurídica': 143,
 

In [ ]:
#Importar nltk
import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')
sw = stopwords.words('portuguese')
nltk.download('punkt')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [ ]:
#Tokenizar dados e remover stopwords

new_sw = (['atender', 'novo', 'contrato', 'saúde', 'coronavírus', 'combate', 'secretaria', 'objeto',
           'covid', 'pública', '2020', 'municipal', 'pi','necessidades', '00', '19', '13', '40',
           'pandemia', 'ações', 'coronavirus', 'município', 'municipio', 'termo', 'emergencial', 'piauí',
           'emergência', 'nº', 'interesse', 'presente', 'decorrente', 'fornecimento', 'valor', 'conforme',
           'aditivo', 'caráter', 'destinados', 'medidas', 'atendimento', 'prevenção', 'necessidade',
           'excepcional', 'lei', 'pacientes', 'enfrentamento', 'corona', 'público', 'situação', 'parnaiba', 
           'junto', 'vírus' ])
sw.extend(new_sw)
#sw.append(['secretaria', 'municipal', 'pi', 'saúde', 'covid']) 
#'prestação', 'contratação', 'aquisição', 'serviços' 

def tokenizar(corpus):
  tokens = []
  for objeto in corpus:
      tokens_objeto = nltk.word_tokenize(objeto, language='portuguese')
      tokens_objeto = [token.lower() for token in tokens_objeto]
      for token in tokens_objeto:
          tokens.append(token)
  tokens_menos_sw = []
  for t in tokens:
    if t not in sw and t.isalpha() :
      tokens_menos_sw.append(t)
  return tokens_menos_sw

In [ ]:
dados_obj = tokenizar(corpus)
dados_obj_simpl = tokenizar(corpus_obj_simp)

dados_totais = dados_obj + dados_obj_simpl

print(len(dados_obj))

78523


### Resultados


Representação binária

In [ ]:
#Binário
import random

#Pegando as palavras mais frequentes
allWords = nltk.FreqDist(w.lower() for w in dados_obj)


#key_words = list(allWords)[0:20]
key_words = ('teste', 'testes', 'rápidos', 'igg', 'igm', 'alimentação', 'reforma', 'medicamentos', 
             'especializada', 'locação', 'serviço', 'serviços', 'cargo', 'epi', 'epis', 'epi\'s', 
             'máscara' 'máscaras', 'materiais', 'empresa', 'luvas', 'leitos', 'limpeza', 'higiene', 
             'hospitalar', 'enfermeiro', 'médico', 'álcool', 'gel', 'insumos', 'veículos', 'equipamentos', 
             'física', 'pessoa', 'temporário', 'função')

def corpus_features(doc):
  i = 1
# doc_words = [w for w in dados_obj]
  doc_words = key_words
  feats = {}
  #if i == 1:
    #print(doc_words)
    #i = i + 1
  for word in doc_words:
    if (word in doc):
      feats[f'countains({word})'] = True
#      feats[f'countains({word})'] = (word in doc)
  return feats

labeled_names = []
for i in range(len(corpus)):
#    print(i, corpus[i], corpus_obj_simp[i])
    labeled_names.append([corpus[i], corpus_obj_simp[i]])  

#labeled_names = ([(name, 'obj') for name in dados_obj[0:5000]] + [(name, 'obj_simpl') for name in dados_obj_simpl[0:5000]])
print(labeled_names[-5:])
random.shuffle(labeled_names)

featuresets = [(corpus_features(n), cat) for (n, cat) in labeled_names]
train_set, test_set = featuresets[0:5400], featuresets[5400:]
classifier = nltk.NaiveBayesClassifier.train(train_set)
print("Taxa de acurácia da classificação: ",nltk.classify.accuracy(classifier, test_set))
print(classifier.show_most_informative_features())

[['Contratação de Empresa para fornecimentos de Teste Rápido para detecção especifica de IGG e IGM do COVID-19, para o Município de Dom Inocêncio - PI.', 'Aquisição de testes para diagnóstico de COVID-19'], ['Prestação de contrataçao de profissional para dar suporte técnico ao desenvolvimento do trabalho de: Acompanhamento a Lei Emergenclal Cultural Aldir Biano, sancionada através do Decreto Federal no 14.017/2020; Alimentação dos sistemas e plataformas necessárias; acompanhamento e julgamento dos editais, todo suporte técnico ao Setor de Cultura junto a Secretaria Municipal de Cultura, Esporte e Lazer e a Prefeitura Municipal de São Jose do Piaul - PI.', 'Contratação de prestador de serviço - pessoa física'], ['AQUISIÇÃO DE PULVERIZADOR 20 LT FIBRA BICO PLASTICO PARA SECRETARIA MUNICIPAL DE EDUCAÇÃO E ESCOLAS DA REDE ESTADUAL E MUNICIPAL DO MUNICÍPIO DE SÃO JOSÉ DO PIAUÍ· PI.', 'Aquisição de equipamentos diversos'], ['AQUISIÇÃO DE ÁGUA SANITÁRIA I LT, DETERGENTE 500 ML E PAPEL TOALHA 

Teste de desempenho (acurácia) - Representação binária

In [ ]:
print("Taxa de acurácia da classificação: ",nltk.classify.accuracy(classifier, test_set))

Taxa de acurácia da classificação:  0.4911392405063291


In [ ]:
import numpy as np
from sklearn.model_selection import KFold

kf = KFold(n_splits=10)

sum_classify = 0
round_classify = 1

for train, test in kf.split(featuresets):
    train_data = np.array(featuresets)[train]
    test_data = np.array(featuresets)[test]
    classifier = nltk.NaiveBayesClassifier.train(train_data)
    print(f"ROUND: {round_classify} - Taxa de acurácia da classificação: ",nltk.classify.accuracy(classifier, test_data))
    sum_classify += nltk.classify.accuracy(classifier, test_data)
    round_classify += 1

average = round(sum_classify/10,3)

print(f"\nTaxa de acurácia média  da classificação: {average}")

ROUND: 1 - Taxa de acurácia da classificação:  0.4491114701130856
ROUND: 2 - Taxa de acurácia da classificação:  0.4911147011308562
ROUND: 3 - Taxa de acurácia da classificação:  0.5137318255250404
ROUND: 4 - Taxa de acurácia da classificação:  0.5088852988691438
ROUND: 5 - Taxa de acurácia da classificação:  0.4620355411954766
ROUND: 6 - Taxa de acurácia da classificação:  0.4878836833602585
ROUND: 7 - Taxa de acurácia da classificação:  0.48465266558966075
ROUND: 8 - Taxa de acurácia da classificação:  0.4911147011308562
ROUND: 9 - Taxa de acurácia da classificação:  0.4927302100161551
ROUND: 10 - Taxa de acurácia da classificação:  0.49434571890145396

Taxa de acurácia média  da classificação: 0.488


Representação Term-Frequency (TF)

In [ ]:
#Formato TF


# Cria o objeto vetorizador
vectorizer_TF = CountVectorizer(analyzer='word', binary=False, decode_error='strict', encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=35, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words=sw,
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, vocabulary=None)


# Aplica o vetorizador nos dados
vectorizer_TF.fit(corpus)


# Saída:

print("Temos um dicionário com o termo e suas respectivas frequências, ou seja quantas vezes aquele termo apareceu na base de dados:\n",vectorizer_TF.vocabulary_)

# Aplicando o transformador
matrix = vectorizer_TF.transform(corpus)

# Visualizando a forma da matriz gerada
print("\nSHAPE:",matrix.shape)

# Imprimindo a matriz binária
print("\n",matrix.toarray())

#esse vetor tem muitos zeros, por essa razão eles são chamados de vetores esparsos.

Temos um dicionário com o termo e suas respectivas frequências, ou seja quantas vezes aquele termo apareceu na base de dados:
 {'empresa': 6, 'prestação': 23, 'serviço': 30, 'contratação': 5, 'pessoa': 20, 'aquisição': 1, 'equipamentos': 8, 'médico': 19, 'risco': 27, 'material': 16, 'hospitalar': 13, 'hospital': 12, 'especializada': 9, 'materiais': 15, 'serviços': 31, 'rede': 26, 'profissionais': 24, 'sanitária': 28, 'medicamentos': 17, 'agente': 0, 'proteção': 25, 'individual': 14, 'testes': 34, 'carga': 3, 'horas': 11, 'semanais': 29, 'máscaras': 18, 'temporário': 33, 'função': 10, 'enfermagem': 7, 'social': 32, 'prazo': 21, 'prestar': 22, 'barreiras': 2, 'cargo': 4}

SHAPE: (6190, 35)

 [[0 0 0 ... 0 0 0]
 [0 1 0 ... 0 0 0]
 [0 1 0 ... 0 0 0]
 ...
 [0 1 0 ... 0 0 0]
 [0 1 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


In [ ]:
def matrix_features(linha):
  feats = {}
  for i in range(len(linha)):
    if linha[i] == 1:
      feats[f'countains({i})'] = True
  return feats


matriz_classif = []
for i in range(len(corpus_obj_simp)):
    matriz_classif.append([matrix.toarray()[i], corpus_obj_simp[i]])  

print(matriz_classif[-5:])
random.shuffle(matriz_classif)

featuresets = [(matrix_features(n), cat) for (n, cat) in matriz_classif]
train_set, test_set = featuresets[0:5400], featuresets[5400:]
classifier_TF = nltk.NaiveBayesClassifier.train(train_set)
print("Taxa de acurácia da classificação: ",nltk.classify.accuracy(classifier_TF, test_set))
print(classifier_TF.show_most_informative_features())

[[array([0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), 'Aquisição de testes para diagnóstico de COVID-19'], [array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), 'Contratação de prestador de serviço - pessoa física'], [array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0]), 'Aquisição de equipamentos diversos'], [array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]), 'Aquisição de material de limpeza'], [array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0]), 'Construção, reforma e/ou ampliação de unidade hospitalar']]
Taxa de acurácia da classificação:  0.6468354430379747
Most Informative Features
           countains(15) = True           Aquisi : Contra =   

Teste de desempenho (acurácia) - Representação Term-Frequency (TF)

In [ ]:
print("Taxa de acurácia da classificação: ",nltk.classify.accuracy(classifier_TF, test_set))

Taxa de acurácia da classificação:  0.6468354430379747


Representação TF-IDF

In [ ]:
#Formato Tf-idf
from sklearn.feature_extraction.text import TfidfVectorizer


# Instanciar o objeto TFIDF 
vectorizer_TF_IDF = TfidfVectorizer(analyzer='word', binary=False, decode_error='strict', encoding='utf-8',
                input='content', lowercase=True, max_df=1.0, max_features=35,
                min_df=1, ngram_range=(1, 1), norm='l2', preprocessor=None,
                smooth_idf=True, stop_words=sw, strip_accents=None,
                sublinear_tf=False, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, use_idf=True, vocabulary=None)


# Constroi o vetorizador nos dados
vectorizer_TF_IDF.fit(corpus)

# saída:

# Imprime o vocabulario
print(vectorizer_TF_IDF.vocabulary_)

print("\n","Essas são as pontuações que foram dadas para cada um dos termos da base de dados. As palavras mais relevantes tem uma pontuação maior:\n",vectorizer_TF_IDF.idf_)

# Aplicando aos dados
matrix = vectorizer_TF_IDF.transform(corpus)

# Imprime a forma da matrix
print("\n","SHAPE:",matrix.shape)
print("\n",matrix.toarray())

{'empresa': 6, 'prestação': 23, 'serviço': 30, 'contratação': 5, 'pessoa': 20, 'aquisição': 1, 'equipamentos': 8, 'médico': 19, 'risco': 27, 'material': 16, 'hospitalar': 13, 'hospital': 12, 'especializada': 9, 'materiais': 15, 'serviços': 31, 'rede': 26, 'profissionais': 24, 'sanitária': 28, 'medicamentos': 17, 'agente': 0, 'proteção': 25, 'individual': 14, 'testes': 34, 'carga': 3, 'horas': 11, 'semanais': 29, 'máscaras': 18, 'temporário': 33, 'função': 10, 'enfermagem': 7, 'social': 32, 'prazo': 21, 'prestar': 22, 'barreiras': 2, 'cargo': 4}

 Essas são as pontuações que foram dadas para cada um dos termos da base de dados. As palavras mais relevantes tem uma pontuação maior:
 [4.2670201  2.20845167 4.43253454 4.44258487 4.44764817 2.26462435
 2.86287749 3.97195013 3.78543129 3.96253091 4.1625074  4.21339901
 3.51624381 4.05752864 4.18177582 3.91374074 3.57387292 3.82821857
 4.37426563 4.21339901 4.25858123 4.37426563 4.41273191 2.5919849
 3.76470516 3.64180703 4.6746061  5.26494378

In [ ]:
def matrix_TF_IDF_features(linha):
  feats = {}
  for i in range(len(linha)):
    if linha[i] != 0:
      if linha[i] >= 0.7:
        feats[f'maior({i})'] = True
      elif linha[i] >= 0.4:
        feats[f'medio({i})'] = True
      else:
        feats[f'menor({i})'] = True
  return feats


matriz_classif = []
for i in range(len(corpus_obj_simp)):
    matriz_classif.append([matrix.toarray()[i], corpus_obj_simp[i]])  

print(matriz_classif[-5:])
random.shuffle(matriz_classif)

featuresets = [(matrix_TF_IDF_features(n), cat) for (n, cat) in matriz_classif]
train_set, test_set = featuresets[0:5400], featuresets[5400:]
classifier_TF_IDF = nltk.NaiveBayesClassifier.train(train_set)
print("Taxa de acurácia da classificação: ",nltk.classify.accuracy(classifier_TF_IDF, test_set))
print(classifier_TF_IDF.show_most_informative_features())

[[array([0.        , 0.        , 0.        , 0.        , 0.        ,
       0.62039634, 0.78428845, 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ]), 'Aquisição de testes para diagnóstico de COVID-19'], [array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0.]), 'Contratação de prestador de serviço - pessoa física'], [array([0.        , 0.42716426, 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.       

### Conclusão

Na questão 1, foi pedido para fazer uma comparação entre os dois modelos de classificadores de gêneros de nomes apresentados em sala de aula. O primeiro modelo apresentava uma única feature, que era a última letra da palavra. O segundo modelo, além da feature anterior, apresentava mais 3 features, a primeira letra do nome, se o nome tinha uma determinada letra e a quantidade vezes que uma determinada letra se petia no nome. O segundo modelo sempre apresentou uma acurácia melhor do que o primeiro modelo, porém sempre com valores muito próximos. Isto ocorreu porque a feature mais representativa era a última letra do nome, sendo, mesmo no modelo 2, 9 das 10 features mais representativas, o que deixou a acurácia dos dois modelos com valores muito próximos. Essa feature diferente que aparecia, "has" ou "count", era o que tornava a acurácia do modelo 2 um pouco melhor. Essa característica continuou quando usamos a técnica 10-fold cross validation. Tanto em cada rodada, como na média final o classificador do modelo 2 sempre apresentou uma acurácia um pouco melhor. Ficando uma média de 0.75 para o modelo 1 e 0.77 para o modelo 2.  

Na questão 2, alteramos nossa função gender_features acrescentado as features "tamanho do nome", "possui ditongo" e "termina com ditongo". Esse modelo apresentou uma acurácia muito parecida com o modelo 2, sendo na maioria das vezes, um valor um pouco superior. E continuou apresentando a "última letra do nome" como feature mais representativa. 

Com relação a questão 3, implementamos classificadores baseados em features com as representações binárias, Term-Frequency (TF) e TF-idF. Utilizamos a planilha utilizada na atividade prática 02 como corpus dessa questão. Essa planilha foi criada com dados de contratospublicados nos diários oficiais do municípios e do Estado do Piauí referentes as compras emergenciais feitas pelos órgãos públicos no período de março a setembro de 2020, durante a pandemia da COVID-19. Dessa planilha foram utilizadas duas colunas como base para o nosso trabalho, "objeto" e "objeto simplificado". O objetivo do modelo era classificar com as classes de "objeto simplificado" o texto da coluna "objeto". Depois de removermos as stopwords, implementamos o modelo de representação e chegamos a uma taxa de acurácia média de 0.35. Acrescentamos às stopwords algumas palavras que tinham uma frequência relativamente alta e não ajudariam na classificação do modelo. Depois de testes na remoção de um conjunto dessas palavras, chegamos a uma taxa de acurácia média de 0.51. Por fim, implementamos os classificadores com as representações TF e TF-idF. Na implementação com TF-idF obtivemos uma taxa de acurácia média de 0,61, enquanto o classificador baseado na representação Term-Frequency, obteve uma taxa de acurácia média de 0,65, nosso melhor resultado. Vale resaltar que por problemas de estouro da RAM do colab, o atributo max_feature nas representações TF e TF-idF foi limitado a 35, obtendo o nosso melhor resultado. Isso demonstra que ainda podemos melhorar essa taxa de acurácia configurando o modelo com o valor do atributo max_feature maior que 35.